### PR TITLE
Support disable metal buffer cache to prevent performance degradation caused by large memory caching

### PR DESCRIPTION
**Pull request by wsvn53**
## Proposed changes

Reason for this PR:

1. When running LLMs inference on devices with smaller memory, such as 8G, the speed noticeably decreases after more and more tokens, and the system performs a large number of Swap memory operations.
2. I observed that when inferring a 1.6B model with a token length of 100, the current consumption in Activity Monitor increased from an initial 3.19GB to around 4GB, and when the token length reached 500, it increased to around 10GB, Even if the reasoning is stopped and all returned results are deleted, the memory usage will not decrease.
image
3. This is not normal because I observed that the fluctuation of GPU memory on 4090 using the same inference code is only a few tens of MB.
4. By looking into the mlx source code, I found that the metal backend of mlx does not release the memory directly when it manages the buffer cache release, and reserves the memory as cache for subsequent tasks to use when allocating memory, but due to a large number of lazy tasks in the inferencing, it accumulates a large number of memory allocations, which leads to the overall memory occupation is too large(maybe).
5. So this change simply adds a getter method and a setter method for `cache_enabled` under the namespace `mlx.core.metal`, in order to maintain compatibility with older versions, it is set to `False` by default.
6. After disabled metal caches by `mlx.core.metal.set_cache_enabled(False)`, even when reasoning with a token length of 1000, the memory growth is only a few tens of MB and the speed does not decrease significantly(~13 tokens/sec on Macmini M1 8GB).

This will make it possible to run coder LLMs or other LLMs by mlx on macOS, supporting longer inference token lengths and becoming more useful.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by awni** [comment]
Cool - I'm ok with this change. But ideally we could fix the underlying cause. The cache really shouldn't be holding on to buffers if the system is under severe memory pressure, so that is unexpected and should be fixed..

I can take a quick look at that before we add this to the API if that's ok. Do you mind telling me the command you were using to observe this behavior? And it was on an 8GB machine correct?

**Comment by wsvn53** [comment]
> Cool - I'm ok with this change. But ideally we could fix the underlying cause. The cache really shouldn't be holding on to buffers if the system is under severe memory pressure, so that is unexpected and should be fixed..
> 
> I can take a quick look at that before we add this to the API if that's ok. Do you mind telling me the command you were using to observe this behavior? And it was on an 8GB machine correct?

Yes, of course you can, just execute the inference command using any model in mlx-examples, and then set max-tokens to a slightly larger value, such as 100-1000. Afterwards, you can observe Memory Usage and Swap Usage through `Activaty Monitor` and `asitop`. With metal cache enabled, the speed becomes slower as more tokens are generated, and even the inference is completed, Memory and Swap Usage remain high. I have reproduced this issue on both Macmini M1 8GB and Macbook Pro M2 16GB.

For example: Qwen-1.8B with fp16 (https://github.com/ml-explore/mlx-examples/tree/main/llms/qwen)：
```bash
python -i qwen.py --max-tokens 500 --prompt 'write a story about The Prince and the Princess.'
```

When the Load Model is completed, the memory usage is approximately 2.68GB, and the overall Swap usage is approximately 6.8/8.0GB:
image
image

After 500 tokens of inference, the memory increased to 5.31GB, and the overall Swap is approximately 8.3/9GB:
image
image

But after `disabling Metal Cache` (Add `mlx.core.metal.set_cache_enabled(False)` to qwen.py), the memory keeps stable even after inferring 1000 tokens (peak memory at 3.9GB):
```bash
python -i qwen.py --max-tokens 1000 --prompt 'write a story about The Prince and the Princess.'
```
image
image
image

Therefore, I think this API helps improve the performance of LLMs in conducting long-context reasoning on Mac.

**Comment by awni** [comment]
I ran some tests, here's the results:

- M2 Mini 8GB
- Command `python qwen.py --max-tokens <500 or 1000>` (using default model Qwen 1B
- Commit 1416e7b664f455aa4ed5092f5faba25d72b1bb17 with and without your patch applied

TLDR 
- It's slightly slower to disable the cache in both cases.
- I don't see any appreciable slow down for longer generations (500 -> 1000).

Cache | Num Tokens | TPS
------ | --------------| ----
No     |  500  | 21
No    | 1000 | 20.2
Yes | 500 | 20.67
Yes | 1000 | 19.8

**Comment by awni** [comment]
It is true that our cache will hold memory even if it's not being used. If MLX is the main application this seems fine, but for not hogging memory with other memory apps, it may be wise to support the option to disable to cache. I can also see it being useful for debugging. 

Idk, I could go either way. Maybe @angeloskath or @jagrit06 have an opinion to sway this.

**Comment by wsvn53** [comment]
> I ran some tests, here's the results:
> 
> * M2 Mini 8GB
> * Command `python qwen.py --max-tokens <500 or 1000>` (using default model Qwen 1B, modified to measure tokens per sec)
> * Commit [1416e7b](https://github.com/ml-explore/mlx/commit/1416e7b664f455aa4ed5092f5faba25d72b1bb17) with and without your patch applied
> 
> TLDR
> 
> * It's slightly slower to disable the cache in both cases.
> * I don't see any appreciable slow down for longer generations (500 -> 1000).
> * Also I did not observe a growth in swap in either setting. @wsvn53, maybe you are using a stale branch?
> 
> Cache	Num Tokens	TPS
> No	500	21
> No	1000	20.2
> Yes	500	20.67
> Yes	1000	19.8

Yes, disabling the cache will slightly reduce token throughput speed. However, I apologize for not mentioning in the previous conversations that there are other applications running on these Mac computers besides mlx. Due to the long-term operation of the system without shutting down, Window Server and Terminal.app like iTeam2 or other necessary open applications also occupy a considerable amount of memory. This will cause significant pressure on mlx-running applications when they increase from 3G to 5G or above. Therefore, as inference progresses, it will become noticeably slower (Mac m1 decreased from 12 tokens/sec to below 5 tokens/sec). So you can observe the memory usage and memory pressure during inferencing.

**Comment by awni** [comment]
@wsvn53 makes sense, I was thinking that that is a good situation where you might want to disable the cache!

Could you rebase and resolve conflicts? I think we can merge this for now, but I might leave it weakly documented until we are comfortable with the API.

**Comment by wsvn53** [comment]
> @wsvn53 makes sense, I was thinking that that is a good situation where you might want to disable the cache!
> 
> Could you rebase and resolve conflicts? I think we can merge this for now, but I might leave it weakly documented until we are comfortable with the API.

Thanks. I have already rebased and resolved the conflicts, please continue to review and merge.

**Comment by wsvn53** [comment]
Sorry, do I need to perform any other actions now in order to continue with the merge? @awni

**Comment by awni** [comment]
Hi @wsvn53 no not at all. My apologies for the delay, I was debating if we should merge this given that we probably intend to remove the API once we fix allocation strategies. But for now I think it's ok to merge (just be warned that this API may possibly be deprecated in the near future).

**Review by awni — APPROVED** [review]
